external-crates/move/crates/move-vm-runtime/src/jit/execution/ast.rs

```diff
@@ -1155,48 +1129,26 @@ impl Type {
 }
 
 impl DatatypeDescriptor {
-    pub fn get_struct(&self) -> PartialVMResult<&StructType> {
-        match self.datatype_info.to_ref() {
-            Datatype::Struct(struct_type) => Ok(struct_type),
-            x @ Datatype::Enum(_) => Err(PartialVMError::new(
-                StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
-            )
-            .with_message(format!("Expected struct type but got {:?}", x))),
-        }
-    }
-
-    pub fn get_enum(&self) -> PartialVMResult<&EnumType> {
-        match self.datatype_info.to_ref() {
-            Datatype::Enum(enum_type) => Ok(enum_type),
-            x @ Datatype::Struct(_) => Err(PartialVMError::new(
-                StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
-            )
-            .with_message(format!("Expected enum type but got {:?}", x))),
-        }
-    }
-
     pub fn datatype_key(&self) -> VirtualTableKey {
-        let module_name = self.module_key;
-        let member_name = self.member_key;
-        VirtualTableKey {
-            package_key: *self.runtime_id.address(),
-            inner_pkg_key: IntraPackageKey {
-                module_name,
-                member_name,
-            },
+        match &self.datatype_info.inner_ref() {
+            Datatype::Enum(vmpointer) => vmpointer.to_ref().def_vtable_key.clone(),
+            Datatype::Struct(vmpointer) => vmpointer.to_ref().def_vtable_key.clone(),
         }
     }
 
     pub fn type_param_constraints(&self) -> impl ExactSizeIterator<Item = &AbilitySet> {
-        self.type_parameters.iter().map(|param| &param.constraints)
+        let type_params = match self.datatype_info.inner_ref() {
+            Datatype::Enum(vmpointer) => &vmpointer.to_ref().type_parameters,
+            Datatype::Struct(vmpointer) => &vmpointer.to_ref().type_parameters,
+        };
```

Comment:
nit: I think this should work here 
```suggestion
        let type_params = self.type_parameters();
```